File: Battleship/src/main/java/de/htwg/battleship/actor/childactor/ShootActor.java
// ShootController.java

package de.htwg.battleship.actor.childactor;

import akka.actor.UntypedActor;
import de.htwg.battleship.actor.messages.ShootMessage;
import de.htwg.battleship.model.IBoard;
import de.htwg.battleship.model.IPlayer;
import de.htwg.battleship.model.IShip;
import de.htwg.battleship.util.StatCollection;

/**
 * ShootController which controlls the shoot method.
 *
 * @author Moritz Sauter (SauterMoritz@gmx.de)
 * @version 1.00
 * @since 2014-11-19
 */
public class ShootActor extends UntypedActor {

    /**
     * Method to shoot on the field of a specific player.
     *
     * @param x      x-Coordinate where to shoot
     * @param y      y-Coordinate where to shoot
     * @param player indicates which player shoots
     *
     * @return true if it was a hit false if not
     */
    public final boolean shoot(final int x, final int y, final IBoard board) {
        if (board.isHit(x, y)) {
            return false;
        } else {
            board.shoot(x, y);
        }
        IShip[] shipList = board.getShipList();
        return hitShipList(shipList, x, y, board.getShips());
    }

    /**
     * Utility-Method if a shoot is a hit or not.
     *
     * @param ship which ship should be test.
     * @param x    x-Coordinate where to shoot
     * @param y    y-Coordinate where to shoot
     *
     * @return true if it is a hit
     */
    private boolean hit(final IShip ship, final int x, final int y) {
        int shipX = ship.getX();
        int shipY = ship.getY();
        int size = ship.getSize();
        if (ship.isOrientation()) {
            int xupp = shipX + size - 1;
            return isHit(xupp, shipX, x, y, shipY);
        } else {
            int yupp = shipY + size - 1;
            //noinspection SuspiciousNameCombination
            return isHit(yupp, shipY, y, x, shipX);
        }
    }

    /**
     * Utility method to check if a ship is hit.
     *
     * @param shipList List of all ships that could be hit
     * @param x        x-Coordinate where to shoot
     * @param y        y-Coordinate where to shoot
     * @param counter  how many ships are on the board
     *
     * @return true if one ship is hit
     */
    private boolean hitShipList(final IShip[] shipList, final int x,
                                final int y, final int counter) {
        for (int i = 0; i < counter; i++) {
            if (hit(shipList[i], x, y)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Utility-Method to combine the Method isBetween and compare the row.
     *
     * @param xupp  upper border
     * @param xlow  lower border
     * @param x     value x
     * @param y     value y
     * @param shipY right row shipY
     *
     * @return true if it is a hit
     */
    private boolean isHit(final int xupp, final int xlow, final int x,
                          final int y, final int shipY) {
        return StatCollection.isBetween(xupp, xlow, x) && y == shipY;
    }

    private IBoard getBoard(final boolean first, final IPlayer player1,
                            final IPlayer player2) {
        return first ? player2.getOwnBoard() : player1.getOwnBoard();
    }

    @Override
    public void onReceive(Object message) throws Exception {
        if (!(message instanceof ShootMessage)) {
            unhandled(message);
            return;
        }
        ShootMessage shootMessage = (ShootMessage) message;
        boolean shootResult = shoot(shootMessage.getX(), shootMessage.getY(),
                                    getBoard(shootMessage.isFirst(),
                                             shootMessage.getPlayer1(),
                                             shootMessage.getPlayer2()));
        getSender().tell(shootResult, getSelf());
    }
}

File: Battleship/src/main/scala/de/htwg/battleship/actor/ShootActor.scala
package de.htwg.battleship.actor

import akka.actor.Actor
import de.htwg.battleship.actor.messages.ShootMessage
import de.htwg.battleship.model.{IBoard, IPlayer, IShip}
import de.htwg.battleship.util.StatCollection

import scala.util.control.Breaks._

/**
  * ShootActor
  *
  * @author ms
  * @since 2016-06-16
  */
object ShootActor {
    val ACTOR_NAME = "shooter"
}

class ShootActor extends Actor {
    def getBoard(first: Boolean, player1: IPlayer, player2: IPlayer): IBoard = {
        if (first) {
            player2.getOwnBoard
        } else {
            player1.getOwnBoard
        }
    }

    def isHit(upper: Int, lower: Int, targetVar: Int, targetFix: Int, fix: Int): Boolean = {
        targetFix == fix && StatCollection.isBetween(upper, lower, targetVar)
    }

    def hit(ship: IShip, x: Int, y: Int): Boolean = {
        val shipX = ship.getX
        val shipY = ship.getY
        val size = ship.getSize
        if (ship.isOrientation) {

            val xupp = shipX + size - 1
            isHit(xupp, shipX, x, y, shipY)
        } else {
            val yupp = shipY + size - 1
            isHit(yupp, shipY, y, x, shipX)
        }
    }

    def hitShipList(shipList: Array[IShip], x: Int, y: Int, counter: Int): Boolean = {
        var result = false
        breakable {
            for (i <- 0 until counter) {
                if (hit(shipList(i), x, y)) {
                    result = true
                    break
                }
            }
        }
        result
    }

    def shoot(x: Int, y: Int, board: IBoard): Boolean = {
        if (board.isHit(x, y)) {
            false
        } else {
            val shipList = board.getShipList
            hitShipList(shipList, x, y, board.getShips)
        }
    }

    override def receive: Receive = {
        case msg: ShootMessage =>
            sender() ! shoot(msg.x, msg.y, getBoard(msg.first, msg.player1, msg.player2))
        case msg => unhandled(msg)
    }
}
